### PR TITLE
 [Refactor]: Validate fix in PR #750 for #759

### LINKED ISSUE #759: [Refactor]: CDAT Migration: `cdms2.open` slice flag adds additional time coordinate which affects downstream time series operations and diffs with Xarray code
### Is your feature request related to a problem?

 In the CDAT version of the Dataset class, cdms2.open is being called with a slice flag (either "co"/"ccb"). This can cause result in major differences between the refactored Xarray code and the CDAT code 

Related lines of code on `main`:
  - https://github.com/E3SM-Project/e3sm_diags/blob/633b52c314325e605fe7f62687cc4d00e5a0a3d5/e3sm_diags/driver/utils/dataset.py#L665-L672
  - https://github.com/E3SM-Project/e3sm_diags/blob/633b52c314325e605fe7f62687cc4d00e5a0a3d5/e3sm_diags/driver/utils/dataset.py#L699-L700
  - https://github.com/CDAT/cdms/blob/3f8c7baa359f428628a666652ecf361764dc7b7a/Lib/axis.py#L392-L414).

In #754, "co" is being set because time coordinates start at the beginning of the month. This adds an additional time coordinate point to the end of the time series file, which affects the subsequent climatology and spatial averaging calculations. I found omitting the extra time coordinate point before calculating the climatology and spatial averaging results in identical results to the xCDAT version.

### Describe the solution you'd like

Should we consider a feature on the slice flag? I found almost no documentation on how to use this slice flag.

  - https://github.com/CDAT/cdms/blob/3f8c7baa359f428628a666652ecf361764dc7b7a/Lib/axis.py#L392-L411

### TODO

- [ ] 1. Find out what "co" and "ccb" do when opening datasets
- [ ] 2. Implement a function that mimics these two cases

### Describe alternatives you've considered

_No response_

### Additional context

- Documentation: https://cdms.readthedocs.io/en/latest/manual/cdms_2.html#axis-methods-additional-to-coordinateaxis-cont-d 
  - [left, right, indicator, cycle] -> e.g,. "co", "ccb"
  - "c" - interval is closed
  - "o" - interval is open
  -  'b' -select axis elements for which the corresponding cell boundary intersects the interval
  - Default is "ccn"
"
- The first 2 letters represents the bounds of the retrievd segment, they can be "c" or "o"
- The third letter represents the search method, it can be "b" ,"n", "e", or "s"
  - The cell will be considered valid if the bounds ("b") or ("n") node are within the interval defined

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Description

- Closes #759 

### 5/30/24
- [x] Re-run `ex1.py` and perform regression test again
- [x] Check for diffs
  - `NET_FLUX_SRF` and `RESTOM` had the largest diffs with the slice flag -- now fixed
- [x] Debug diffs
	- [x] Fix "TREFHT" land large diff with mean metrics -- due to incorrect Xarray datasets being passed to `_subset_on_region()` (need to pass objects with land mask applied)
	- [x] Fix `TREFHT` land mean ref regridded data (11.59% diff) -- **address this later in a separate GH issue (#817)**

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: auxiliary_tools/cdat_regression_testing/759-slice-flag/debug/trefht.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+mv2_reg_filled = mv2_reg.filled(np.nan)
+
+#  Count of np.nan: 22170 vs. 22170 (same)
+np.count_nonzero(np.isnan(mv2_filled.flatten())) == np.count_nonzero(
+    np.isnan(mv2_reg_filled.flatten())
+)
+# Sum: -68837.01023016105 vs. -68837.01023016105 (same)
+np.nansum(mv2_filled) == np.nansum(mv2_reg_filled)
+# Mean: -6.342086809486 == -6.342086809486 (same)
+np.nanmean(mv2_filled) == np.nanmean(mv2_reg_filled)
```
Comment (tomvothecoder) [inline auxiliary_tools/cdat_regression_testing/759-slice-flag/debug/trefht.py:68]: I noticed that for the "TREFHT" variable, the CDAT code on `main` reaches the conditional to regrid the reference variable to the test variable. However, the regridded reference data is the same as the original reference data as noted in this comparison.

https://github.com/E3SM-Project/e3sm_diags/blob/c8404ab0f5523fee80cda396357f307b7ad1ff5b/e3sm_diags/driver/utils/general.py#L246-L254

File: auxiliary_tools/cdat_regression_testing/759-slice-flag/debug/trefht.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+# Count of np.nan: 23150 vs. 22170 (different)
+np.count_nonzero(np.isnan(ds_b_regrid[var_key].values.flatten())) == np.count_nonzero(
+    np.isnan(mv2_reg_filled.flatten())
+)
+# Sum: -77674.06388742107 vs. -68837.01023016105 (different)
+np.nansum(ds_b_regrid[var_key].values) == np.nansum(mv2_reg_filled.data)
+# Mean: -7.866524598685545 vs. -6.342086809486 (different)
+np.nanmean(ds_b_regrid[var_key].values) == np.nanmean(mv2_reg_filled.data)
+
+# %%
```
Comment (tomvothecoder) [inline auxiliary_tools/cdat_regression_testing/759-slice-flag/debug/trefht.py:93]: I noted how xCDAT actually performs the regridding. In [debug_TREFHT_diff.png](https://github.com/E3SM-Project/e3sm_diags/pull/815/files#diff-30ca25d6805e404f6be184ca614b9c48996f024860c546b7a1a7507cc381c97e) we can see the difference in the plots. I think there are missing values along the coastlines with xCDAT?

**Review by tomvothecoder — COMMENTED** [review]
This PR validates the replication of the slice flag feature added in PR #750 to fix #759. 

I re-ran `ex1.py` and found that all `.nc` files produced are now within the relative tolerance. For metrics, all but 1 metric is within the relative tolerance.

This 1 metrics is related to the regridded reference data for `"TREFHT"` with the `land` region. It seems like xCDAT is properly regridding the variable using xESMF, while the CDAT code does nothing with ESMF. 

I think we should move forward for now and possible debug this more later on.

File: e3sm_diags/driver/utils/regrid.py [hunk 3]
```diff
@@ -91,8 +91,8 @@ def subset_and_align_datasets(
 
     # Subset on a specific region.
     if "global" not in region:
-        ds_test_new = _subset_on_region(ds_test, var_key, region)
-        ds_ref_new = _subset_on_region(ds_ref, var_key, region)
+        ds_test_new = _subset_on_region(ds_test_new, var_key, region)
+        ds_ref_new = _subset_on_region(ds_ref_new, var_key, region)
```
Comment (tomvothecoder) [inline e3sm_diags/driver/utils/regrid.py:95]: Incorrect dataset objects were being passed to `_subset_on_region`. If land sea mask was applied, it was being ignored here and downstream.

This affected `lat_lon` and `polar`.

**Comment by tomvothecoder** [comment]
Hey @chengzhuzhang, just an FYI that I added functionality to replicate the `cdms2` `"ccb` flag for non-submonthly data (#759).

I tested `ex1.py` which uses time series data and all of the results check out except one (new issue #817). You can refer to the above review [comment](https://github.com/E3SM-Project/e3sm_diags/pull/815#pullrequestreview-2094829305) for more info.

**Comment by chengzhuzhang** [comment]
@tomvothecoder great work to support the slice flag and match the results with main! Also thanks for careful documentation on the final straggler and I agree that we should address in a separate PR..

**Pull request merged**